Frameworks/CoreGraphics/DWriteWrapper.mm

```diff
@@ -580,29 +637,238 @@ HRESULT STDMETHODCALLTYPE CreateStreamFromKey(_In_ const void* fontFileReference
+        CFRetain(fontDatas);
+        m_fontDatas.reset(fontDatas);
+        m_previouslyCreatedFiles = previouslyCreatedFiles;
+        return S_OK;
+    }
+
+    HRESULT STDMETHODCALLTYPE GetCurrentFontFile(_Out_ IDWriteFontFile** fontFile) {
+        RETURN_HR_IF(E_ILLEGAL_METHOD_CALL, m_location < 0 || m_location > CFArrayGetCount(m_fontDatas.get()));
+
+        if (0 <= m_location && m_location < m_previouslyCreatedFiles->size()) {
+            *fontFile = m_previouslyCreatedFiles->at(m_location).Get();
+        } else {
+            ComPtr<IDWriteFactory> dwriteFactory;
+            RETURN_IF_FAILED(DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, __uuidof(IDWriteFactory), &dwriteFactory));
+
+            ComPtr<DWriteFontBinaryDataLoader> loader;
+            RETURN_IF_FAILED(
+                MakeAndInitialize<DWriteFontBinaryDataLoader>(&loader, (CFDataRef)CFArrayGetValueAtIndex(m_fontDatas.get(), m_location)));
+            RETURN_IF_FAILED(dwriteFactory->RegisterFontFileLoader(loader.Get()));
+
+            int unused;
+            RETURN_IF_FAILED(dwriteFactory->CreateCustomFontFileReference(&unused, sizeof(unused), loader.Get(), fontFile));
+            m_previouslyCreatedFiles->emplace_back(*fontFile);
+        }
+
+        return S_OK;
+    }
+
+    HRESULT STDMETHODCALLTYPE MoveNext(_Out_ BOOL* hasCurrentFile) {
+        *hasCurrentFile = ++m_location < CFArrayGetCount(m_fontDatas.get());
+        return S_OK;
+    }
+
+private:
+    int m_location = -1;
+
+    woc::unique_cf<CFArrayRef> m_fontDatas;
+    std::vector<ComPtr<IDWriteFontFile>>* m_previouslyCreatedFiles;
 };
 
 /**
+ * Implementation of IDWriteFontCollectionLoader, which creates an IDWriteFontFileEnumerator to create font files
+ * Which are used to create a font collection.  This is expected to be a singleton
+ */
+class DWriteFontCollectionLoader : public RuntimeClass<RuntimeClassFlags<WinRtClassicComMix>, IDWriteFontCollectionLoader> {
+protected:
+    InspectableClass(L"Windows.Bridge.DirectWrite", TrustLevel::BaseTrust);
+
+public:
+    DWriteFontCollectionLoader() {
+    }
+
+    HRESULT RuntimeClassInitialize() {
+        m_fontDatas.reset(CFArrayCreateMutable(kCFAllocatorDefault, 0, &kCFTypeArrayCallBacks));
+        m_fontDatasSet.reset(CFSetCreateMutable(nullptr, 0, &kCFTypeSetCallBacks));
+        return S_OK;
+    }
+
+    void AddDatas(CFArrayRef fontDatas, CFArrayRef* errors) {
+        CFMutableArrayRef outErrors = nil;
+        if (errors) {
+            outErrors = CFArrayCreateMutable(nullptr, 0, &kCFTypeArrayCallBacks);
+            *errors = outErrors;
+        }
+
+        CFIndex count = CFArrayGetCount(fontDatas);
+        for (CFIndex i = 0; i < count; ++i) {
+            CFDataRef data = (CFDataRef)CFArrayGetValueAtIndex(fontDatas, i);
+            if (data) {
+                if (CFSetContainsValue(m_fontDatasSet.get(), data)) {
+                    if (errors) {
```

Comment:
this amount of nesting is pretty hard to read - consider making a helper that creates & appends an error, if !outErrors, and an overload for appending nullptr